### PR TITLE
Wismy/racing mode

### PR DESCRIPTION
**Pull request by WismyYao**

### CODE REVIEW

File: radio/src/gui/128x64/model_setup.cpp [hunk 1]
```diff
@@ -1230,6 +1232,16 @@ void menuModelSetup(event_t event)
         break;
       }
 
+#if defined(HARDWARE_INTERNAL_MODULE)
+      case ITEM_MODEL_SETUP_INTERNAL_MODULE_RACING_MODE:
+      {
+        ModuleData & moduleData = g_model.moduleData[moduleIdx];
+        lcdDrawTextAlignedLeft(y, STR_RACING_MODE);
+        moduleState[moduleIdx].racingMode = editCheckBox(moduleState[moduleIdx].racingMode, MODEL_SETUP_2ND_COLUMN, y, "", attr, event);
```
Comment (bsongis) [inline radio/src/gui/128x64/model_setup.cpp]: it won't be saved in the model storage, it means that on watchdog reset we will be in an unknown state

File: radio/src/gui/212x64/model_setup.cpp [hunk 2]
```diff
@@ -1098,6 +1100,15 @@ void menuModelSetup(event_t event)
         break;
       }
 
+      case ITEM_MODEL_SETUP_INTERNAL_MODULE_RACING_MODE:
+      {
+        uint8_t moduleIdx = CURRENT_MODULE_EDITED(k);
+        ModuleData & moduleData = g_model.moduleData[moduleIdx];
+        lcdDrawTextAlignedLeft(y, STR_RACING_MODE);
+        moduleState[moduleIdx].racingMode = editCheckBox(moduleState[moduleIdx].racingMode, MODEL_SETUP_2ND_COLUMN, y, "", attr, event);
```
Comment (bsongis) [inline radio/src/gui/212x64/model_setup.cpp]: same comment here, the variable should be persistent

File: radio/src/gui/gui_common.h [hunk 3]
```diff
@@ -184,6 +184,26 @@ inline uint8_t MODULE_CHANNELS_ROWS(int moduleIdx)
   }
 }
 
+inline bool isRacingModeAllowed()
+{
+  if (isModulePXX2(INTERNAL_MODULE) && 0 == g_model.moduleData[INTERNAL_MODULE].channelsCount && moduleState[INTERNAL_MODULE].racingMode)
```
Comment (bsongis) [inline radio/src/gui/gui_common.h]: expression can be simplified

File: radio/src/gui/gui_common.h [hunk 4]
```diff
@@ -184,6 +184,26 @@ inline uint8_t MODULE_CHANNELS_ROWS(int moduleIdx)
...
+  else
+    return false;
+}
+
+inline uint8_t IF_ALLOW_RACING_MODE(int moduleIdx)
+{
+  if (!IS_MODULE_ENABLED(moduleIdx)) {
+    return HIDDEN_ROW;
+  }
+  else if (isModulePXX2(moduleIdx)) {
```
Comment (bsongis) [inline radio/src/gui/gui_common.h]: `else if (isModulePXX2(moduleIdx) && 0 == g_model.moduleData[moduleIdx].channelsCount)`

File: radio/src/gui/gui_common.h [hunk 5]
```diff
@@ -184,6 +184,26 @@ inline uint8_t MODULE_CHANNELS_ROWS(int moduleIdx)
...
+    if (0 == g_model.moduleData[moduleIdx].channelsCount)
+      return 1;
```
Comment (bsongis) [inline radio/src/gui/gui_common.h]: return 1 means 2 fields, do we have 2 fields or only one? in this case it should be 0

File: radio/src/pulses/pulses.cpp [hunk 6]
```diff
@@ -526,10 +526,11 @@ void setCustomFailsafe(uint8_t moduleIndex)
         g_model.failsafeChannels[ch] = channelOutputs[ch];
       }
     }
+    storageDirty(EE_MODEL);
```
Comment (bsongis) [inline radio/src/pulses/pulses.cpp:640]: good catch

File: radio/src/pulses/pxx.h [hunk 7]
```diff
@@ -29,7 +29,7 @@
 
 #define PXX2_LOWSPEED_BAUDRATE             230400
 #define PXX2_HIGHSPEED_BAUDRATE            450000
-#define PXX2_PERIOD                        4 // 4ms
+#define PXX2_PERIOD                        50 // 4ms
```
Comment (bsongis) [inline radio/src/pulses/pxx.h]: why 50?

File: radio/src/pulses/pxx2.cpp [hunk 8]
```diff
@@ -47,14 +47,20 @@ void Pxx2Pulses::addFlag1(uint8_t module)
   else {
     subType = g_model.moduleData[module].subType;
   }
-  Pxx2Transport::addByte(subType << 4);
+
+  uint8_t flag1 = subType << 4u;
+  if (isFunctionActive(FUNCTION_RACING_MODE)) {
```
Comment (bsongis) [inline radio/src/pulses/pxx2.cpp]: + test for "allowed"!

File: radio/src/translations/pt.h.txt [hunk 9]
```diff
@@ -1244,7 +1245,7 @@
 #define TR_JITTER_FILTER       "ADC Filter"
 #define TR_RTC_CHECK           TR("Check RTC", "Check RTC voltage")
 #define TR_AUTH_FAILURE        "Auth-failure"
-
+#define TR_RACING_MODE         INDENT "Racing mode"
```
Comment (bsongis) [inline radio/src/translations/pt.h.txt]: don't introduce INDENT there, we try to remove them slowly

File: radio/src/gui/480x272/model_setup.cpp [hunk 10]
```diff
@@ -1247,6 +1249,17 @@ bool menuModelSetup(event_t event)
         }
         break;
       }
+
+#if defined(HARDWARE_INTERNAL_MODULE)
+      case ITEM_MODEL_SETUP_INTERNAL_MODULE_RACING_MODE:
+      {
+        ModuleData & moduleData = g_model.moduleData[moduleIdx];
```
Comment (bsongis) [inline radio/src/gui/480x272/model_setup.cpp:1256]: moduleData is not used anywhere!

**Comment by bsongis** [comment]
The Companion part is missing

File: radio/src/pulses/pxx.h (hunk 7 above)
Comment (WismyYao) [reply radio/src/pulses/pxx.h]: You suggested me change it to 50ms . Racing mode is 4.6ms , we should set it more then 4.6ms

Comment (3djc) [reply radio/src/pulses/pxx.h]: 50 doesn't look right at all to me. It is also used on external module btw

Comment (bsongis) [reply radio/src/pulses/pxx.h]: right, I have pushed a proposal, it needs testing (internal module + heartbeat)

**Pull request merged**